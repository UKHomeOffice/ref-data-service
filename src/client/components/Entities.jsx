import React from 'react';
import { Link } from 'react-router-dom';

// local imports
import Banner from 'Banner';
import config from '../../config/core';

const {apiUrls} = config;

const EntitiesData = ({ data }) => {
  return data.map((entities) => {
    return (
      <tr className="govuk-table__row" key={entities.id}>
        <th className="govuk-table__header" scope="row">{entities.label}</th>
        <td className="govuk-table__cell">{entities.description}</td>
        <td className="govuk-table__cell"><Link to={`/entities/${entities.entityName}`}>View data</Link></td>
        <td className="govuk-table__cell"><Link to={`/entities/${entities.entityName}/schema`}>View definition</Link></td>
      </tr>
    )
  })
};

export default class Entities extends React.Component {
  constructor(props) {
    super(props);
    this.state = {
      entitiesObject: {}
    };
  }

  componentDidMount() {
    const entities = apiUrls.entities;
    fetch(entities)
      .then(res => res.json())
      .then(obj => {
        this.setState({ entitiesObject: obj })
      });
  }

  render() {
    return (
      <div className="govuk-width-container">
        <Banner/>
        <main id="main-content" className="govuk-main-wrapper " role="main">
          <div className="govuk-grid-row">
            <div className="govuk-grid-column-two-thirds">
              <h1 className="govuk-heading-xl">Reference Data Governance Tool</h1>
              <p className="govuk-body-l">This service allows you to view and manage reference data entities.</p>
              <h2 className="govuk-heading-l">Data Entities</h2>
              <span></span>
              <table className="govuk-table">
                <thead className="govuk-table__head">
                  <tr className="govuk-table__row">
                    <th className="govuk-table__header" scope="col">Name</th>
                    <th className="govuk-table__header" scope="col">Description</th>
                    <th className="govuk-table__header" scope="col">Data Items</th>
                    <th className="govuk-table__header" scope="col">Entity Definition</th>
                  </tr>
                </thead>
                <tbody className="govuk-table__body">
                  {this.state.entitiesObject && this.state.entitiesObject.data &&
                    <EntitiesData data={this.state.entitiesObject.data} />
                  }
                </tbody>
              </table>
            </div>
          </div>
        </main>
      </div>
    );
  }
}

