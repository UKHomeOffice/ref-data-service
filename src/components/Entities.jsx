import React from 'react';
import { Link } from 'react-router-dom';
import { connect } from 'react-redux';

// local imports
import Banner from './Banner';
import ServiceUnavailable from './ServiceUnavailable';
import config from '../../config/core';
import logger from '../../logger';
import { compareEntities } from '../utils';

const {apiUrls, serviceDesk} = config;

const EntitiesData = ({ data }) => {
  const entityRows = []

  data.map((entity, n) => {
    let { entityName } = entity;
    let { description, label } = entity.schema;

    entityRows.push(
      <tr className="govuk-table__row" key={n}>
        <th className="govuk-table__header" scope="row">{label}</th>
        <td className="govuk-table__cell">{description}</td>
        <td className="govuk-table__cell">
          <Link to={`/entities/${entityName}`}>View</Link>

          { config.readOnly ?
            <React.Fragment></React.Fragment>
          :
            <React.Fragment>
              <Link to={`/entities/${entityName}/schema/edit/description`}>Edit</Link>
              <Link to={`/entities/${entityName}/delete`}>Delete</Link>
            </React.Fragment>
          }
        </td>
      </tr>
    )
  })
  return entityRows;
};

class Entities extends React.Component {
  constructor(props) {
    super(props);
    this.state = { entitiesObject: {}, error: null };
  }

  componentDidMount() {
    const entities = apiUrls.entities;

    fetch(entities, {
      method: 'GET',
      headers: {
        'Accept': 'application/json',
        'Authorization': `Bearer ${this.props.kc.token}`,
      }
    })
    .then(res => {
      if (res.status !== 200) {
        logger.error(`Error status: ${res.status}, message: ${res.statusText}`);
        throw Error(res.statusText);
      }
      return res.json();
    })
    .then(obj => {
      obj.data = obj.data.sort(compareEntities);
      this.setState({ entitiesObject: obj })
    })
    .catch(error => this.setState({ error }));
  }

  render() {
    if (this.state.error) {
      return <ServiceUnavailable />
    }

    return (
      <div className="govuk-width-container">
        <Banner/>
        <main id="main-content" className="govuk-main-wrapper " role="main">
          <div className="govuk-grid-row">
            <div className="govuk-grid-column-two-thirds">
              <h1 className="govuk-heading-xl">Reference Data Service</h1>
              <p className="govuk-body">This service allows you to view and manage reference data sets. To add a new data set to the repository, use the <a className='govuk-anchor-click' href={`${serviceDesk.addDataSet}`}>Add a data set</a> link at the top of the page.</p>
              <h2 className="govuk-heading-l">Data Sets</h2>
              <span></span>
              <table className="govuk-table entities-table">
                <thead className="govuk-table__head">
                  <tr className="govuk-table__row">
                    <th className="govuk-table__header" scope="col">Name</th>
                    <th className="govuk-table__header" scope="col">Description</th>
                    <th className="govuk-table__header" scope="col">Action</th>
                  </tr>
                </thead>
                <tbody className="govuk-table__body">
                  {this.state.entitiesObject && this.state.entitiesObject.data &&
                    <EntitiesData data={this.state.entitiesObject.data} />
                  }
                </tbody>
              </table>
            </div>
          </div>
        </main>
      </div>
    );
  }
}

const mapStateToProps = state => ({
 'kc': state.keycloak
});

export default connect(mapStateToProps)(Entities);
