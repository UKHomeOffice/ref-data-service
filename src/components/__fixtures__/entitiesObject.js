const entitiesObject = {
  'status': 'success',
  'code': 200,
  'data': [
    {
      'entityName': 'country',
      'schema': {
        'label': 'Countries',
        'description': 'Countries',
        'schemalastupdated': '10/03/2019',
        'dataversion': 1,
        'required': [
          'id',
          'iso31661alpha2',
          'iso31661alpha3',
          'name',
          'continent',
          'iso31661numeric'
        ],
        'properties': {
          'id': {
            'format': 'integer',
            'type': 'integer',
            'description': {
              'label': 'Identifier',
              'description': 'database unique identity record',
              'summaryview': 'false'
            }
          },
          'iso31661alpha2': {
            'maxLength': 2,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': '2 digit alpha code',
              'description': 'Country 2 Character alpha code',
              'summaryview': 'true'
            }
          },
          'iso31661alpha3': {
            'maxLength': 3,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': '3 digit alpha code',
              'description': 'Country 3 Character alpha code',
              'summaryview': 'true'
            }
          },
          'name': {
            'maxLength': 40,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': 'Country name',
              'description': 'Country name',
              'summaryview': 'true'
            }
          },
          'continent': {
            'maxLength': 2,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': 'Continent',
              'description': 'Countinent country is part of',
              'summaryview': 'true'
            }
          },
          'dial': {
            'maxLength': 20,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': 'Phone dial code',
              'description': 'Country dailing prefix',
              'summaryview': 'true'
            }
          },
          'iso31661numeric': {
            'format': 'integer',
            'type': 'integer',
            'description': {
              'label': '3 digit numeric code',
              'description': 'Country numeric ISO code',
              'summaryview': 'true'
            }
          },
          'validfrom': {
            'format': 'date',
            'type': 'string',
            'description': {
              'label': 'Valid from date',
              'description': 'Item valid from date',
              'summaryview': 'false'
            }
          },
          'validto': {
            'format': 'date',
            'type': 'string',
            'description': {
              'label': 'Valid to date',
              'description': 'Item valid to date',
              'summaryview': 'false'
            }
          }
        }
      },
    },
    {
      'entityName': 'nationality',
      'label': 'Nationalities',
      'description': 'The list of recognised Nationalities, including some additonal information.',
      'schema': {
        'description': {
          'description': 'Nationalities',
          'schemalastupdated': '10/03/2019',
          'dataversion': 1
        },
        'required': [
          'id',
          'nationality',
          'visarequired',
          'evwoptional',
          'diplomaticexception',
          'specialexception'
        ],
        'properties': {
          'id': {
            'format': 'integer',
            'type': 'integer',
            'description': {
              'label': 'Identifier',
              'description': 'database unique identity record',
              'summaryview': 'false'
            }
          },
          'nationality': {
            'maxLength': 330,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': 'Identifier',
              'description': 'Nationality names',
              'summaryview': 'true'
            }
          },
          'iso31661alpha3': {
            'maxLength': 3,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': '3 digit alpha code',
              'description': 'Country 3 Character alpha code',
              'summaryview': 'true'
            }
          },
          'iso31661alpha2': {
            'maxLength': 2,
            'format': 'character varying',
            'type': 'string',
            'description': {
              'label': '2 digit alpha code',
              'description': 'Country 2 Character alpha code',
              'summaryview': 'true'
            }
          },
          'visarequired': {
            'format': 'boolean',
            'type': 'boolean',
            'description': {
              'label': 'Visa required',
              'description': 'Is VISA required to visit UK',
              'summaryview': 'false'
            }
          },
          'evwoptional': {
            'format': 'boolean',
            'type': 'boolean',
            'description': {
              'label': 'Optional - EVW',
              'description': 'Is Electronic Visa Waver optional to visit UK',
              'summaryview': 'false'
            }
          },
          'diplomaticexception': {
            'format': 'boolean',
            'type': 'boolean',
            'description': {
              'label': 'Exception - Diplomatic',
              'description': 'Are there diplomatic exceptions for visiting the UK',
              'summaryview': 'false'
            }
          },
          'specialexception': {
            'format': 'boolean',
            'type': 'boolean',
            'description': {
              'label': 'Exception - Special',
              'description': 'Are there special exceptions for visiting the UK',
              'summaryview': 'false'
            }
          },
          'countryid': {
            'format': 'integer',
            'type': 'integer',
            'description': {
              'label': 'Linked country id',
              'description': 'Country link to Country dataset',
              'summaryview': 'false',
              'linkedrecord': 'country(id)'
            }
          },
          'validfrom': {
            'format': 'date',
            'type': 'string',
            'description': {
              'label': 'Valid from date',
              'description': 'Item valid from date',
              'summaryview': 'false'
            }
          },
          'validto': {
            'format': 'date',
            'type': 'string',
            'description': {
              'label': 'Valid to date',
              'description': 'Item valid to date',
              'summaryview': 'false'
            }
          }
        }
      }
    }
  ]
};

export default entitiesObject;
