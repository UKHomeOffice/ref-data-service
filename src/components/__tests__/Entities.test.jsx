import React from 'react';
import configureMockStore from 'redux-mock-store';
import { shallow } from 'enzyme';

// local imports
import Entities from '../Entities';
import entitiesObject from '../__fixtures__/entitiesObject';

const mockStore = configureMockStore();

describe('Entities component', () => {
  const keycloakToken = {
    'keycloak': {
      'token': '1357',
      'tokenParsed': {
        'name': 'Eve Polastri',
        'email': 'eve.polastri@mail.com',
      },
    },
  };
  
  beforeEach(() => {
    global.fetch = jest
      .fn()
      .mockImplementation(() => Promise.resolve({
        'json': () => entitiesObject
      }));    
  });

  it('should render a table with the right data', () => {
    const store = mockStore(keycloakToken);
    const wrapper = shallow(<Entities store={ store }/>).dive();

    wrapper.setState({
      'entitiesObject': entitiesObject
    });

    expect(wrapper.find('EntitiesData').dive()).toMatchSnapshot();
  });

  it('Should display the entities component successfully', () => {    
    const store = mockStore(keycloakToken);
    const wrapper = shallow(<Entities store={ store }/>).dive();

    wrapper.setState({
      'entitiesObject': entitiesObject
    });

    // headers
    expect(wrapper.containsMatchingElement(<h1 className="govuk-heading-xl">Reference Data Service</h1>)).toBeTruthy();
    expect(wrapper.containsMatchingElement(<h2 className="govuk-heading-l">Data Sets</h2>)).toBeTruthy();

    // table headers
    expect(wrapper.find('.govuk-table__row').children()).toHaveLength(3);
    expect(wrapper.find('.govuk-table__header').get(0).props.children).toEqual('Name');
    expect(wrapper.find('.govuk-table__header').get(1).props.children).toEqual('Description');
    expect(wrapper.find('.govuk-table__header').get(2).props.children).toEqual('Action');

    // table content
    expect(wrapper.find('EntitiesData').dive().find('.govuk-table__row').children()).toHaveLength(6);
    for (let i = 0; i < entitiesObject.data.length; i++) {
      expect(wrapper.find('EntitiesData').dive().get(i).props.children[0].props.children).toEqual(entitiesObject.data[i].schema.label);
      expect(wrapper.find('EntitiesData').dive().get(i).props.children[1].props.children).toEqual(entitiesObject.data[i].schema.description);            
      expect(wrapper.find('EntitiesData').dive().get(i).props.children[2].props.children[0].props.children).toEqual('View');
      expect(wrapper.find('EntitiesData').dive().get(i).props.children[2].props.children[1].props.children[0].props.children).toEqual('Edit');
      expect(wrapper.find('EntitiesData').dive().get(i).props.children[2].props.children[1].props.children[1].props.children).toEqual('Delete');
    };
  });

  afterEach(() => {
    jest.resetAllMocks();
  });
});



