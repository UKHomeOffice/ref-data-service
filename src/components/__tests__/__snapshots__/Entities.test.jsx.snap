// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Entities component should render a table with the right data 1`] = `
Array [
  <tr
    className="govuk-table__row"
    key="0"
  >
    <th
      className="govuk-table__header"
      scope="row"
    >
      Countries
    </th>
    <td
      className="govuk-table__cell"
    >
      Countries
    </td>
    <td
      className="govuk-table__cell"
    >
      <Link
        replace={false}
        to="/entities/country"
      >
        View
      </Link>
      <Link
        replace={false}
        to="/entities/country/schema/edit/description"
      >
        Edit
      </Link>
      <Link
        replace={false}
        to="/entities/country/delete"
      >
        Delete
      </Link>
    </td>
  </tr>,
  <tr
    className="govuk-table__row"
    key="1"
  >
    <th
      className="govuk-table__header"
      scope="row"
    />
    <td
      className="govuk-table__cell"
    >
      <Component />
    </td>
    <td
      className="govuk-table__cell"
    >
      <Link
        replace={false}
        to="/entities/nationality"
      >
        View
      </Link>
      <Link
        replace={false}
        to="/entities/nationality/schema/edit/description"
      >
        Edit
      </Link>
      <Link
        replace={false}
        to="/entities/nationality/delete"
      >
        Delete
      </Link>
    </td>
  </tr>,
]
`;
